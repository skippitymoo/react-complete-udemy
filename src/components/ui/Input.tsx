import React, { JSX, useId } from 'react';

type InputProps = React.InputHTMLAttributes<HTMLInputElement> & {
  /**
   * The label for the input
   */
  label: string;
};

export const Input = ({ label, className, ...rest }: InputProps): JSX.Element => {
  const id = useId();
  className = className ?? '';

  return (
    <div className='form__group'>
      <label htmlFor={id} className='form__label'>
        {label}
      </label>
      <input {...rest} id={id} className={`form__input ${className}`} />
    </div>
  );
};
