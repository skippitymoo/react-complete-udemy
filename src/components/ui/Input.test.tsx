import { render, screen } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import { Input } from './Input';
import userEvent from '@testing-library/user-event';

expect.extend(toHaveNoViolations);

describe('Input', () => {
  it('renders all the props', async () => {
    const { container } = render(<Input label='label yeo' className='yeo-class' />);

    expect(await axe(container)).toHaveNoViolations();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('handles input', async () => {
    const handleChangeFn = jest.fn();
    render(<Input label='Name' onChange={(e) => handleChangeFn(e.target.value)} />);

    const input: HTMLInputElement = screen.getByLabelText('Name');
    expect(input.value).toBe('');

    await userEvent.type(input, 'Joe Casual');

    expect(input.value).toBe('Joe Casual');
    expect(input).not.toBeInvalid();
    expect(handleChangeFn).toBeCalledWith('Joe Casual');
  });
});
