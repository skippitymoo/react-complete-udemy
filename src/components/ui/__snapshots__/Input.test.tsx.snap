// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders all the props 1`] = `
<div
  class="form__group"
>
  <label
    class="form__label"
    for=":r0:"
  >
    label yeo
  </label>
  <input
    class="form__input yeo-class"
    id=":r0:"
  />
</div>
`;
